### PR TITLE
Introduce List feature.

### PR DESCRIPTION
**Pull request by scofalik**
Introduces List feature. ckeditor/ckeditor5#2936 

Related PR: https://github.com/ckeditor/ckeditor5-engine/pull/616

### CODE REVIEW

**Comment by scofalik** [comment]
Some insights:
1. The indenting is partially implemented but it is buggy at the moment and will be done later in separate issue (#5). It will probably happen next iteration.
2. I don't really like `src/utils.js` and I will gladly take any advice where to move them (and whether the functions should be a little bit generalised).

**Comment by scofalik** [comment]
Related PR: https://github.com/ckeditor/ckeditor5-engine/pull/616

File: src/listengine.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		data.viewToModel.on( 'element:li', viewModelConverter );
+		data.viewToModel.on( 'element:ul', cleanList, { priority: 'high' } );
+		data.viewToModel.on( 'element:ol', cleanList, { priority: 'high' } );
+
+		// Register commands for numbered and bulleted list.
+		editor.commands.set( 'numberedList', new ListCommand( editor, 'numbered' ) );
+		editor.commands.set( 'bulletedList', new ListCommand( editor, 'bulleted' ) );
+
+		// Register commands for indenting.
+		editor.commands.set( 'indentList', new IndentCommand( editor, 'forward' ) );
```
Comment (Reinmar) [inline src/listengine.js:88]: Shouldn't `indentListItem` and `outdentListItem` sound nicer?

**Comment by Reinmar** [comment]
What can I say? The list feature works amazingly well. I mean, really, I weren't expecting it to be so stable and cover such cases from the day one. Great job!

Check this:

sep-28-2016 15-09-19

Check also this super tricky case, when <kbd>backspace</kbd> should merge an item which is outside the selection before the merge happens (the "Another bullet list" item). It works!

sep-28-2016 15-27-28

And what I love the most is that we were able to introduce the list feature in a way that it's compatible with other features (deleting, enter, headings), without touching the other features. What's more – there is actually only one listener in the list feature code which integrates it with the <kbd>enter</kbd> key. In CKEditor 4 there were couple thousand lines of code spread across a big chunk of the code base which glued all the things together. Impressive, how we managed to design it in CKEditor 5 :).

File: src/list.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+				data.preventDefault();
+				evt.stop();
+			}
+		} );
+
+		// Add tab key support.
+		// When in list item, pressing tab should indent list item, if possible.
+		// Pressing shift + tab shout outdent list item.
+		this.listenTo( this.editor.editing.view, 'keydown', ( evt, data ) => {
```
Comment (Reinmar) [inline src/list.js:56]: Can't we do this through the keystroke handler?

**Review by Reinmar — APPROVED** [review]
I left two questions in the code. I'm not sure about the names (I'm now more for the shorter ones) and I think that we can use keystroke handler for the keys (but I now that this is discussed in https://github.com/ckeditor/ckeditor5-core/issues/20, because it concerns more features).

So... I think we're good to go here.

**Pull request merged**